cmds/exp/pxeserver/main.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// pxeserver is a test & lab PXE server that supports TFTP, HTTP, and DHCPv4.
+//
+// pxeserver can either respond to *all* DHCP requests, or a DHCP request from
+// a specific MAC. In either case, it will supply the same IP in all answers.
+package main
+
+import (
+	"bytes"
+	"flag"
+	"log"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/insomniacslk/dhcp/dhcpv4"
+	"github.com/insomniacslk/dhcp/dhcpv4/server4"
+	"pack.ag/tftp"
+)
+
+var (
+	mac = flag.String("mac", "", "MAC address to respond to. Responds to all requests if unspecified.")
+
+	// DHCPv4-specific
+	selfIP       = flag.String("ip", "192.168.0.1", "DHCPv4 IP of self")
+	yourIP       = flag.String("your-ip", "192.168.0.2/24", "The one and only CIDR to give to all DHCPv4 clients")
+	rootpath     = flag.String("rootpath", "", "RootPath option to serve via DHCPv4")
+	bootfilename = flag.String("bootfilename", "pxelinux.0", "Boot file to serve via DHCPv4")
+	inf          = flag.String("interface", "eth0", "Interface to serve DHCPv4 on")
+
+	// File serving
+	tftpDir = flag.String("tftp-dir", "", "Directory to serve over TFTP")
+	httpDir = flag.String("http-dir", "", "Directory to serve over HTTP")
+)
+
+type server struct {
+	mac          net.HardwareAddr
+	yourIP       net.IP
+	submask      net.IPMask
+	self         net.IP
+	bootfilename string
+	rootpath     string
+}
+
+func (s *server) dhcpHandler(conn net.PacketConn, peer net.Addr, m *dhcpv4.DHCPv4) {
+	log.Printf("Handling request %v", m)
+
+	var replyType dhcpv4.MessageType
+	switch mt := m.MessageType(); mt {
+	case dhcpv4.MessageTypeDiscover:
+		replyType = dhcpv4.MessageTypeOffer
+	case dhcpv4.MessageTypeRequest:
+		replyType = dhcpv4.MessageTypeAck
+	default:
+		log.Printf("Can't handle type %v", mt)
+		return
+	}
+	if s.mac != nil && !bytes.Equal(m.ClientHWAddr, s.mac) {
+		log.Printf("Not responding to DHCP request for mac %s, which does not match %s", m.ClientHWAddr, s.mac)
```

Comment:
at some future date would you want this to be an error channel that some top level thing watches?